### DISCUSSION TITLE
Packages aren't in the environment when using nested lists of targets?

### DISCUSSION THREAD

**Discussion by tarensanders** [ropensci/targets#644] [Help] [answered]
Hi folks,

In the [manual](https://books.ropensci.org/targets/walkthrough.html#target-script-file) it says:

> Every target script must end with a list of your tar_target() objects. **Those objects can be nested, i.e. lists within lists.**

I interpreted this to mean that I can create lists of targets elsewhere, and then pass them to one final list in the `_targets.R` script.

But when I do this, the functions do not seem to have access to the globally loaded packages, even when I set them with `tar_option_set`. Here's an example (where `%>%` is not found).

``` r
library(targets)

tar_dir({

  lines <- "
  targetlist <- 
  list(
    tar_target(mtdata,
               data(mtcars)),
    tar_target(pipedtarget,
              myfunc(mtdata)))
  
  myfunc <- function(df){
  df %>% head()
  }
  
  "
  
  writeLines(lines, "source.R")

  tar_script({
    source("source.R")
    
    # I'm explicitly passing packages that have "%>%"
    packages <- c("tidyverse", "dplyr")
    tar_option_set(packages = packages)
    
    # The nested list
    list(
      targetlist
    )
    
    
  })
  tar_manifest()
  tar_make()


})
#> * start target mtdata
#> * built target mtdata
#> * start target pipedtarget
#> x error target pipedtarget
#> * end pipeline
#> Error : could not find function "%>%"
#> Error: callr subprocess failed: could not find function "%>%"
#> Visit https://books.ropensci.org/targets/debugging.html for debugging advice.
```

<sup>Created on 2021-09-27 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

If I run debug on this (following the instructions [here](https://books.ropensci.org/targets/debugging.html#the-debug-option)), I find that there are no packages loaded in the environment. What am I doing wrong?

**Edit**: Just to add that the results of the debugging of `pipedtarget` look like this:

```
Browse[1]> (.packages())
[1] "targets"   "stats"     "graphics"  "grDevices" "datasets"  "utils"     "methods"   "base"  
```

I'm not clear on why `dplyr` or `tidyverse` aren't being passed.

**Comment by wlandau** [comment, marked as answer]
`tar_option_set()` sets the default arguments to `tar_target()`. So it will work if you call `tar_option_set()` before `source("source.R")`.

**Reply by tarensanders** [reply]
Yikes, that's an obvious mistake in retrospect. Thank you!

Thanks for this package BTW. It would be hard to overstate how much it's transformed our workflow.
